packages/core/src/utils/GroupGraph.ts

```diff
@@ -0,0 +1,62 @@
+import { Shape, ShapeAD } from "../types/shape";
+import Graph from "./Graph";
+
+export type GroupGraph = Graph<
+  string, // shape name
+  number // shape layer
+>;
+
+export const makeGroupGraph = <T extends Shape | ShapeAD>(
+  shapes: T[]
+): GroupGraph => {
+  const graph: GroupGraph = new Graph();
+  const nameShapeMap: { [k: string]: T } = {};
```

Comment:
Since you're using this as a map, can you use [`Map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map)?

```suggestion
  const nameShapeMap = new Map<string, ShapeAD>();
```

Reply:
Done